powerauth-credential-server-sample/src/main/java/io/getlime/security/powerauth/lib/credentials/controller/SMSAuthorizationController.java

```diff
@@ -0,0 +1,167 @@
+import org.springframework.stereotype.Controller;
+import org.springframework.validation.BeanPropertyBindingResult;
+import org.springframework.validation.ValidationUtils;
+import org.springframework.web.bind.MethodArgumentNotValidException;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RequestMethod;
+import org.springframework.web.bind.annotation.ResponseBody;
+
+import java.math.BigDecimal;
+import java.security.SecureRandom;
+import java.util.Date;
+import java.util.UUID;
+
+/**
+ * Controller class which handles SMS OTP authorization.
+ *
+ * @author Roman Strobl, roman.strobl@lime-company.eu
+ */
+@Controller
+public class SMSAuthorizationController {
+
+    private SMSAuthorizationRepository smsAuthorizationRepository;
+    private CredentialStoreConfiguration credentialStoreConfiguration;
+
+    public SMSAuthorizationController(SMSAuthorizationRepository smsAuthorizationRepository, CredentialStoreConfiguration credentialStoreConfiguration) {
+        this.smsAuthorizationRepository = smsAuthorizationRepository;
+        this.credentialStoreConfiguration = credentialStoreConfiguration;
+    }
+
+    /**
+     * Create a new SMS OTP authorization message.
+     *
+     * @param request Request data.
+     * @return Response with message ID.
+     */
+    @RequestMapping(value = "/sms/create", method = RequestMethod.POST)
+    public @ResponseBody
+    ObjectResponse<CreateSMSAuthorizationResponse> create(@RequestBody ObjectRequest<CreateSMSAuthorizationRequest> request) throws MethodArgumentNotValidException {
+        CreateSMSAuthorizationRequest createSMSAuthorizationRequest = request.getRequestObject();
+
+        // input validation is handled by CreateSMSAuthorizationRequestValidator
+        // validation is invoked manually because of the generified Request object
+        CreateSMSAuthorizationRequestValidator validator = new CreateSMSAuthorizationRequestValidator();
+        BeanPropertyBindingResult result = new BeanPropertyBindingResult(createSMSAuthorizationRequest, "createSMSAuthorizationRequest");
+        ValidationUtils.invokeValidator(validator, createSMSAuthorizationRequest, result);
+        if (result.hasErrors()) {
+            // getEnclosingMethod() on new object returns a reference to current method
+            MethodParameter methodParam = new MethodParameter(new Object() {
+            }.getClass().getEnclosingMethod(), 0);
+            throw new MethodArgumentNotValidException(methodParam, result);
+        }
+
+        // messageId is generated as random UUID, it can be overriden to provide a real message identification
+        String messageId = UUID.randomUUID().toString();
+
+        // authorization code generation - to be updated based on application requirements
+        String authorizationCode = String.valueOf((new SecureRandom()).nextInt(90000000) + 10000000);
```

Comment:
This is not correct implementation. We are not generating random authorization code. We need to keep the property of "dynamic linking" in place. Basically, we need to make sure that if amount and account number changes, the SMS OTP changes. One way to achieve this is to implement following:

```java
byte[] key = Generator.generateRandomBytes(16);
byte[] signatureBase = amount + "&" + currency + "&" account;
byte[] hash = Mac.hmacSha256(signatureBase, key);
String otp = Util.decimalize(hash);
```

Then, we need to store "key"

You can use PowerAuth crypto, to make your life easier...

Please have a look at:

- https://github.com/lime-company/lime-security-powerauth/blob/ea3c6863b78817e977ed80dc3ce228d3bd8c052e/powerauth-java-crypto/src/main/java/io/getlime/security/powerauth/crypto/lib/util/HMACHashUtilities.java#L42
- https://github.com/lime-company/lime-security-powerauth/blob/ea3c6863b78817e977ed80dc3ce228d3bd8c052e/powerauth-java-crypto/src/main/java/io/getlime/security/powerauth/crypto/lib/util/SignatureUtils.java#L118
- https://github.com/lime-company/lime-security-powerauth/blob/ea3c6863b78817e977ed80dc3ce228d3bd8c052e/powerauth-java-crypto/src/main/java/io/getlime/security/powerauth/crypto/lib/generator/KeyGenerator.java#L108